### PR TITLE
[FLINK-9642]Reduce the count to deal with state during a CEP process

### PR DESCRIPTION
**Pull request by Aitozi**
## What is the purpose of the change

With the rework of sharedBuffer Flink-9418, the lock & release operation is deal with state backend which is different from the previous version which will read the state of sharedBuffer all to memory, i think we can add a cache or variable in sharedbuffer to cache the LockAble Object to mark the ref change in once process in NFA, this will reduce the count when the events point to the same Node. And flush the result to MapState at the end of process. 

## Brief change log

- add the eventsBufferCache and entryCache
- flush the cache after one turn process

### CODE REVIEW

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 1]
```diff
@@ -346,16 +351,87 @@ private void lockEvent(EventId eventId) throws Exception {
...
 
+	// Cache related method
+
+	/////////////////////////////////////////////
+	//  Put
+	/////////////////////////////////////////////
+
+	/**
+	 * Put a event to cache.
+	 * @param eventId id of the event
```
Comment (zhangminglei) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Put an event.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 2]
```diff
@@ -346,16 +351,87 @@ private void lockEvent(EventId eventId) throws Exception {
...
+		this.entryCache.put(nodeId, entry);
+	}
+
+	/////////////////////////////////////////////
+	// Get
+	/////////////////////////////////////////////
+
+	/**
+	 * Try to get the sharedBufferNode from state iff the node has not been quered during this turn process.
+	 * @param nodeId id of the event
```
Comment (zhangminglei) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: iff ? I think it should be if.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 3]
```diff
@@ -75,6 +76,9 @@
 	private MapState<Long, Integer> eventsCount;
 	private MapState<NodeId, Lockable<SharedBufferNode>> entries;
 
+	private HashMap<EventId, Lockable<V>> eventsBufferCache = new HashMap<>();
+	private HashMap<NodeId, Lockable<SharedBufferNode>> entryCache = new HashMap<>();
```
Comment (zhangminglei) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: I would suggest change it to 

`private Map<EventId, Lockable<V>> eventsBufferCache = new HashMap<>();`

**Comment by zhangminglei** [comment]
Hi, @Aitozi Could you please take a look on the travis error 

```
Tests run: 2, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.014 sec <<< FAILURE! - in org.apache.flink.cep.nfa.NFAStateAccessTest
testIterativeWithABACPattern(org.apache.flink.cep.nfa.NFAStateAccessTest)  Time elapsed: 0.009 sec  <<< FAILURE!
java.lang.AssertionError: expected:<20> but was:<8>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.apache.flink.cep.nfa.NFAStateAccessTest.testIterativeWithABACPattern(NFAStateAccessTest.java:193)

testComplexBranchingAfterZeroOrMore(org.apache.flink.cep.nfa.NFAStateAccessTest)  Time elapsed: 0.005 sec  <<< FAILURE!
java.lang.AssertionError: expected:<5> but was:<2>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.apache.flink.cep.nfa.NFAStateAccessTest.testComplexBranchingAfterZeroOrMore(NFAStateAccessTest.java:112)

Running org.apache.flink.cep.nfa.compiler.NFACompilerTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.019 sec - in org.apache.flink.cep.nfa.compiler.NFACompilerTest
Running org.apache.flink.cep.nfa.DeweyNumberTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.01 sec - in org.apache.flink.cep.nfa.DeweyNumberTest
Running org.apache.flink.cep.nfa.sharedbuffer.SharedBufferTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec - in org.apache.flink.cep.nfa.sharedbuffer.SharedBufferTest
Running org.apache.flink.cep.pattern.PatternTest
Tests run: 22, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.033 sec - in org.apache.flink.cep.pattern.PatternTest

Results :

Failed tests: 
  NFAStateAccessTest.testComplexBranchingAfterZeroOrMore:112 expected:<5> but was:<2>
  NFAStateAccessTest.testIterativeWithABACPattern:193 expected:<20> but was:<8>

Tests run: 80, Failures: 2, Errors: 0, Skipped: 12

13:21:32.852 [INFO] ------------------------------------------------------------------------
13:21:32.852 [INFO] Reactor Summary:
13:21:32.852 [INFO] 
13:21:32.852 [INFO] flink-queryable-state-client-java .................. SUCCESS [  4.830 s]
13:21:32.852 [INFO] flink-cep .......................................... FAILURE [  5.861 s]
13:21:32.852 [INFO] flink-table ........................................ SKIPPED
13:21:32.852 [INFO] flink-queryable-state-runtime ...................... SKIPPED
13:21:32.852 [INFO] flink-gelly ........................................ SKIPPED
13:21:32.852 [INFO] flink-gelly-scala .................................. SKIPPED
13:21:32.852 [INFO] flink-gelly-examples ............................... SKIPPED
13:21:32.852 [INFO] flink-python ....................................... SKIPPED
13:21:32.852 [INFO] flink-ml ........................................... SKIPPED
13:21:32.852 [INFO] flink-cep-scala .................................... SKIPPED
```

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java (hunk 2 above)
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: it means `if and only if`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java (hunk 3 above)
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: agree and fixed

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java (hunk 2 above)
Comment (zhangminglei) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Yes. Thanks let me know this.

**Comment by Aitozi** [comment]
Hi @zhangminglei ,  thanks for your review. I only check the SharedBufferTest locally before, the error in travis means the num of state access (read and write) is less than before which is the purpose of this pr, and I fix the error. cc @dawidwys

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java (hunk 1 above)
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: fixed

**Comment by Aitozi** [comment]
The travis error this time seems unrelated.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 4]
```diff
@@ -346,16 +351,87 @@ private void lockEvent(EventId eventId) throws Exception {
...
+		Lockable<V> event = eventsBufferCache.get(eventId);
+		return event != null ? event : eventsBuffer.get(eventId);
+	}
+
+	/**
+	 * Flush the event and node in map to state.
+	 * @throws Exception Thrown if the system cannot access the state.
+	 */
+	public void flushCache() throws Exception {
+		entryCache.forEach((k, v) -> {
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Ok, Is this benefit from the `RocksDBWriteBatchWrapper` when use the `putAll`?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 5]
```diff
@@ -346,16 +351,87 @@ private void lockEvent(EventId eventId) throws Exception {
...
+					throw new RuntimeException();
+				}
+			}
+		);
+
+		eventsBufferCache.forEach((k, v) -> {
+				try {
+					eventsBuffer.put(k, v);
+				} catch (Exception e) {
+					throw new RuntimeException();
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: But I don't know how to deal with the exception in a stream api in java8, do you have a better way to deal with this situation? thanks.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sihuazhou) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Yes.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sihuazhou) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: In fact, what I means is that if you want to throw an exception here, you could throw the exception as `throw new RuntimeException("exception message", originalException)`, this way the original exception won't be swallowed.

[the remark this answers is not in the document]
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Get it.

**Comment by Aitozi** [comment]
Using the `entries#putAll` in `flushCache` lead to the count in `NFAStateAccessTest` increased,  I will check it locally , this travis will fail.

**Comment by Aitozi** [comment]
fixed the error that the access to state increased in `NFAStateAccessTest` by add the `isEmpty` judgment before update the state.

**Comment by Aitozi** [comment]
Is this  look ok now? ping @sihuazhou @dawidwys

**Comment by Aitozi** [comment]
Could you take a look at this PR @dawidwys ?

**Comment by zhangminglei** [comment]
Hi,  @Aitozi  I think it has conflicts now.

**Comment by Aitozi** [comment]
Resolved the conflicts,  please help review when you free @dawidwys .

**Comment by dawidwys** [comment]
Hi @Aitozi , 
Thanks for the work here. I think we could improve a bit separation of concerns in the SharedBuffer. I am a bit afraid this class will become too complex once again. How about we split the SharedBuffer into two layers: caching one(SharedBuffer) and accessing buffer e.g. `SharedBufferAccessor`.  We could make the accessor `AutoCloseable`. I think it would give more explicit information about the necessity to flush the buffer.

We could move to `SharedBufferAccessor` methods like:
- registerEvent
- put
- extractPatterns
- materializeMatch
- materializeMatch
- lockNode
- releaseNode
- removeNode
- lockEvent
- releaseEvent

In the `SharedBuffer` we would leave only the caching package-protected methods. We would also add a method there to get the `SharedBufferAccessor` that would `flush` the changes on `close`. This way we would have a cleaner separation, plus we would make the flushing more intuitive.

What do you think?

**Comment by dawidwys** [comment]
Hi @Aitozi , 
Thanks for the work here. I think we could improve a bit separation of concerns in the SharedBuffer. I am a bit afraid this class will become too complex once again. How about we split the SharedBuffer into two layers: caching one(SharedBuffer) and accessing buffer e.g. `SharedBufferAccessor`.  We could make the accessor `AutoCloseable`. I think it would give more explicit information about the necessity to flush the buffer.

We could move to `SharedBufferAccessor` methods like:
- registerEvent
- put
- extractPatterns
- materializeMatch
- materializeMatch
- lockNode
- releaseNode
- removeNode
- lockEvent
- releaseEvent

In the `SharedBuffer` we would leave only the caching package-protected methods. We would also add a method there to get the `SharedBufferAccessor` that would `flush` the changes on `close`. This way we would have a cleaner separation, plus we would make the flushing more intuitive.

What do you think?

**Comment by Aitozi** [comment]
A little busy these day and delayed this PR, Now I have separate `SharedBuffer` to 
- `SharedBuffer` to deal with the NFA query 
- `SharedBufferAccessor` to deal with the underlay `State` directly

Please take a look again @dawidwys ,thx.

**Comment by dawidwys** [comment]
Hi @Aitozi 
I found the dependencies between accessor and shared buffer a bit too much both sided. Please have a look at my idea how would I loosen it a bit in this branch: https://github.com/dawidwys/flink/tree/pr/6205. It does not compile, but you should get the idea how would I split up those two classes.

**Comment by Aitozi** [comment]
Thanks a lot for your review👍. I have chaned as your mentioned, please take a look then, thx @dawidwys

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 6]
```diff
@@ -204,227 +133,78 @@ public NodeId put(
...
-								edge.getDeweyNumber(),
-								newPath));
-						}
-					}
-				}
-
-			}
-		}
-		return result;
+	void cacheEvent(EventId eventId, Lockable<V> event) {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Rename to `upsertEvent`

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 7]
```diff
@@ -204,227 +133,78 @@ public NodeId put(
...
+		Lockable<V> event = eventsBufferCache.get(eventId);
+		return event != null ? event : eventsBuffer.get(eventId);
 	}
 
 	/**
-	 * Decreases the reference counter for the given event so that it can be
-	 * removed once the reference counter reaches 0.
-	 *
-	 * @param eventId id of the event
+	 * Flush the event and node from cache to state.
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Missing empty line

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 8]
```diff
@@ -204,227 +133,78 @@ public NodeId put(
...
-	private void lockEvent(EventId eventId) throws Exception {
-		Lockable<V> eventWrapper = eventsBuffer.get(eventId);
-		checkState(
-			eventWrapper != null,
-			"Referring to non existent event with id %s",
-			eventId);
-		eventWrapper.lock();
-		eventsBuffer.put(eventId, eventWrapper);
+	Lockable<V> getEvent(EventId eventId) throws Exception {
+		Lockable<V> event = eventsBufferCache.get(eventId);
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Same as above

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 9]
```diff
@@ -204,227 +133,78 @@ public NodeId put(
...
-	public void releaseNode(final NodeId node) throws Exception {
-		Lockable<SharedBufferNode> sharedBufferNode = entries.get(node);
-		if (sharedBufferNode != null) {
-			if (sharedBufferNode.release()) {
-				removeNode(node, sharedBufferNode.getElement());
-			} else {
-				entries.put(node, sharedBufferNode);
-			}
-		}
+	Lockable<SharedBufferNode> getEntry(NodeId nodeId) throws Exception {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: Replace with: 

	Lockable<SharedBufferNode> getEntry(NodeId nodeId) {
		return entryCache.computeIfAbsent(nodeId, id -> {
			try {
				return entries.get(id);
			} catch (Exception ex) {
				throw new WrappingRuntimeException(ex);
			}
		};
	}

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java [hunk 10]
```diff
@@ -161,40 +100,30 @@ public void init(
...
-			final EventId eventId,
-			@Nullable final NodeId previousNodeId,
-			final DeweyNumber version) throws Exception {
+	public SharedBufferAccessor<V> getAccessor() {
+		return new SharedBufferAccessor<>(this);
+	}
 
-		if (previousNodeId != null) {
-			lockNode(previousNodeId);
+	public void advanceTime(long timestamp) throws Exception {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: This could be package protected.

Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/sharedbuffer/SharedBuffer.java]: This need to be called in nfa package, I think it should be public.

**Comment by Aitozi** [comment]
Have fixed the place you pointed, and add a test to cover the element in state and cache's change. Please take a look again, thx. @dawidwys

**Comment by Aitozi** [comment]
The failed travis ci seems unrelated with this PR.

**Comment by Aitozi** [comment]
Hi, @dawidwys  can you help review again, 3Q ?

**Comment by dawidwys** [comment]
Sorry for slow response, I will try to have a final look next week, but from what I've seen so far it looks quite well.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 11]
```diff
@@ -216,9 +217,12 @@ private boolean isFinalState(ComputationState state) {
 			final T event,
 			final long timestamp,
 			final AfterMatchSkipStrategy afterMatchSkipStrategy) throws Exception {
-		try (EventWrapper eventWrapper = new EventWrapper(event, timestamp, sharedBuffer)) {
-			return doProcess(sharedBuffer, nfaState, eventWrapper, afterMatchSkipStrategy);
+		Collection<Map<String, List<T>>> result;
+		try (SharedBufferAccessor<T> sharedBufferAccessor = sharedBuffer.getAccessor();
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: OK, i will modify the parameter of `advanceTime` and another `process` as well to make the `NFA` only work with `SharedBufferAccessor`

**Comment by Aitozi** [comment]
Have made the NFA only works with `sharedBufferAccessor`, ping @dawidwys

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATestUtilities.java [hunk 12]
```diff
@@ -66,22 +67,25 @@
...
-				inputEvent.getValue(),
-				inputEvent.getTimestamp(),
-				afterMatchSkipStrategy);
-
-			for (Map<String, List<Event>> p: patterns) {
-				List<Event> res = new ArrayList<>();
-				for (List<Event> le: p.values()) {
-					res.addAll(le);
+
+		try (SharedBufferAccessor<Event> sharedBufferAccessor = sharedBuffer.getAccessor()) {
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATestUtilities.java]: Fixed. Thanks for your carefully review 👍

**Comment by Aitozi** [comment]
ping @dawidwys

**Pull request merged**